### DISCUSSION TITLE
Jenkinsfile parsing/testing/manipulating - Does it work?

### DISCUSSION THREAD

**Discussion by crankydillo** [openrewrite/rewrite#4559] [Q&A] [answered]
I need to modify some Jenkinsfile and started out with a test similar to this and impl code that uses your `TreeVisitingPrinter`:

```java
rewriteRun(
                //language=groovy
                groovy(
                        """
                    @Library('hello')_
                    somePipeline {
                    }
                          """,
                        """
                    @Library('hello')_
                    somePipeline {
                      defaults = someDefaults()
                    }     
                          """,
                        spec -> spec.path("Jenkinsfile")
                )
        );
```

When I run this via JUnit, I get:

```
java.lang.AssertionError: Source file was parsed into an LST that contains non-whitespace characters in its whitespace. This is indicative of a bug in the parser. 
@Li~~(non-whitespace)~~>brary('hello')<~~_
somePipeline {
}

	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:323)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:131)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:126)
```

I assume this is happening because it's using your `GroovyParser`, and that understandably doesn't handle this Jenkins DSL thing (I think there's a question somewhere about this)..  However, when I use my recipe via the MavenPlugin, the tree printer does seem to walk the statements.  So 2 questions.

1. Can I reasonably modify the Jenkinsfile even with these test problems?  I'm about to try that out, but figured I'd ask anyway:). Maybe it will help some others.  I know I've spoken with people who think OpenRewrite can easily modify Jenkinsfiles in the same way it can modify Maven pom files and Java source code.
2. Is there any non-hacky way to make the test work?  Can I reasonably bypass the whitespace validation?  I'm going down a pretty hacky override path of `GroovyBuilder`, where I'm overriding `service`...

**Comment by crankydillo** [comment]
Currently, I'm omitting Jenkinsfile DSL stuff like `@Library("foo")_` (pretty sure this is Jenkinsfile DSL, not Groovy) and asserting on changes that can be successfully represented as Groovy.  Then I do some adhoc testing on real Jenkinsfile.  Seems to be working ok.  Ideally, I could test the `Jenkinsfile`.  Maybe I could try treating is a text or something, but somehow force the use of my visitor.

**Reply by shanman190** [reply]
So I would have expected a `Jenkinsfile` to parse without any issues. Based upon the output there, it seems the Groovy parser visitor got confused on something...

**Reply by timtebeek** [reply]
Hmm; we do have a similar if slightly different issue report already:
- https://github.com/openrewrite/rewrite/issues/3824

Both need a closer look at the parser to see this resolved, and the tests are a helpful first step, thanks! Did you already explore a possible fix as well?

**Reply by crankydillo** [reply]
Sorry for the late reply @timtebeek , but no.  I have not had time to explore a fix.  Unfortunately, this is causing another issue where we need to modify the @Library 'statement'.  Some of that relates to #4827.  But, I'm just getting back from a break:). Need to ramp up on where we are.

**Reply by timtebeek** [reply]
All good; we've just put out a new version of OpenRewrite with more Groovy parser fixes, so best to try with that version:
https://github.com/openrewrite/rewrite/releases/tag/v8.42.5

**Reply by crankydillo** [reply]
A quick test suggests that has not fixed this for us, but all I did was manage rewrite-groovy to 8.42.5.  Still waiting for the recipe-bom release:P

Initially, this was just a test thing, but now we actually have to change that `@Library` annotation.  I think we're trying to use a text-regex hack to work around this problem.  I plan to create an issue soon.

**Reply by crankydillo** [reply]
So what it's looking like is

```
@Library('hello')_
```

roughly equivalent to the first 2 lines of this:
```
@Deprecated
someStr

someStr = "abc"
````

The `_` is basically an 'unnamed' variable declaration, which plays some role in how jenkins pipelines work.  [This code](https://github.com/openrewrite/rewrite/blob/7444d1c3321b487d940894afecc1200822b0d1e8/rewrite-groovy/src/main/java/org/openrewrite/groovy/GroovyParserVisitor.java#L1474) doesn't seem to handle annotations on variable declarations at all (note the `emptyList`).  It somehow mishandles while trying to determine a type/identifier..  I'll continue trying to write tests, but if you see something obviously wrong with this claim, it could help me avoid some unnecessary work:)

**Reply by crankydillo** [reply, marked as answer]
- Created https://github.com/openrewrite/rewrite/issues/4853

I do wish github would link these like it does with issues/prs.
